dfget/core/downloader/p2p_downloader/client_writer.go

```diff
@@ -246,6 +246,9 @@ func writePieceToFile(piece *Piece, file *os.File, cdnSource apiTypes.CdnSource)
 
 	writer := pool.AcquireWriter(file)
 	_, err := io.Copy(writer, piece.RawContent(noWrapper))
+	if piece.autoReset {
+		piece.ResetContent()
+	}
```

Comment:
You can not ResetContent here in writePieceToFile.
Because the same piece may be used in target writer queue after write finish in client writer queue. 
https://github.com/dragonflyoss/Dragonfly/blob/master/dfget/core/downloader/p2p_downloader/client_writer.go#L223

Reply:
when enable cross write, we will read piece content twice? but this is a buffer, it has index, when read twice, I think we need reset some index, #1411 is a good workaround.

Reply:
> when enable cross write, we will read piece content twice? but this is a buffer, it has index, when read twice, I think we need reset some index, #1411 is a good workaround.

#1411 still can not solve this problem. Underlying Buf data will be reused. Piece should be reset after all writer process.

Reply:
I have rewritten the fix code #1411 , could you help me review whether it can work?

Reply:
_great!_